doc/doxygen/code-gallery/CMakeLists.txt

```diff
@@ -0,0 +1,89 @@
+## ---------------------------------------------------------------------
+##
+## Copyright (C) 2015 by the deal.II Authors
+##
+## This file is part of the deal.II library.
+##
+## The deal.II library is free software; you can use it, redistribute
+## it, and/or modify it under the terms of the GNU Lesser General
+## Public License as published by the Free Software Foundation; either
+## version 2.1 of the License, or (at your option) any later version.
+## The full text of the license can be found in the file LICENSE at
+## the top level of the deal.II distribution.
+##
+## ---------------------------------------------------------------------
+
+
+
+#
+# A target for the preparation of all the stuff happening in here...
+#
+
+ADD_CUSTOM_TARGET(code-gallery)
+
+
+#
+# Define ways to get the code-gallery repository downloaded at
+# the time that we call cmake, so that we can subsequently
+# use the files that we find during download as inputs for
+# doxygen. This has to happen at the time we run cmake because
+# doing so at build time is too late for us to set up the
+# proper dependencies for the build targets.
+#
+# To do so, we need to execute git on a command line; this
+# (and consequently everything else below) will not work
+# on Windows. It will also not work if we can't find a version
+# of git
+#
+FIND_PACKAGE(Git)
+IF( (NOT CMAKE_SYSTEM_NAME MATCHES "CYGWIN")
+    AND
+    (NOT CMAKE_SYSTEM_NAME MATCHES "Windows")
+    AND GIT_FOUND )
+
+  # First update or download the code-gallery repository
+  IF (IS_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/code-gallery)
+    MESSAGE(STATUS "Updating code-gallery git repository")
+    EXECUTE_PROCESS(COMMAND ${GIT_EXECUTABLE} checkout master
+                    COMMAND git pull origin master
+                    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/code-gallery
+                    TIMEOUT 30
+                    OUTPUT_QUIET
+                    ERROR_QUIET
+		    RESULT_VARIABLE _result)
+  ELSE()
+    MESSAGE(STATUS "Cloning code-gallery git repository")
+    EXECUTE_PROCESS(COMMAND ${GIT_EXECUTABLE} clone https://github.com/bangerth/code-gallery.git
+                    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+                    TIMEOUT 30
+                    OUTPUT_QUIET
+                    ERROR_QUIET
+		    RESULT_VARIABLE _result)
+  ENDIF()
```

Comment:
I'm strongly against any download/update of a repository during configure/make. It is broken in many ways. For example
  - What if the ```git pull``` leads to a merge conflict? Then, suddenly you have to know that sources are stored in a subdirectory below the build directory.
  - What if I want to create documentation of a fresh change (not yet committed) to the code gallery? This is not possible with this setup.

The only solution for this is to have ```code-gallery``` (similarly to ```examples```) as a top level directory in the sources.

Therefore, I would suggest that we figure out a way how to deal with subprojects in general (that can be freely added/removed) from the sources and are picked up dynamically.

git submodules is a way to go. It would have the advantage of also being suitable for ```bundled``` and ```examples```. It records a blessed subproject commit in the main repository (and synchronizes history with that). The big disadvantage is the fact that it is not exactly easy to work with them...

Alternatively, we could also go with something that does not couple history that tightly.